### PR TITLE
Snapshot: don't walk directories without read permissions

### PR DESCRIPTION
**Pull request by BoboTiG**
Original patch by Joshua Skelton (@joshuaskelly) on issue #408.

### CODE REVIEW

**Comment by BoboTiG** [comment]
@rrzaripov, do you mind reviewing :pray: ?

File: src/watchdog/utils/dirsnapshot.py [hunk 1]
```diff
@@ -220,7 +220,7 @@ def walk(root):
                 # list of its parent and trying to delete its contents. If this
                 # happens we treat it as empty. Likewise if the directory was replaced
                 # with a file of the same name (less likely, but possible).
-                if e.errno == errno.ENOENT or e.errno == errno.ENOTDIR:
+                if e.errno in (errno.ENOENT, errno.ENOTDIR, errno.EINVAL):
```
Comment (rrzaripov) [inline src/watchdog/utils/dirsnapshot.py]: Looks like this was be merged in #559?

**Comment by rrzaripov** [comment]
@BoboTiG I was tested this patch under Windows.
For example, next directory structure exists c:\tmp\A\B\C.  All directories for simplifying empty. While testing we will disabling access to B.

Python 2.7
```python
>>> from watchdog.utils import dirsnapshot
>>> dirsnapshot.DirectorySnapshot("c:\\tmp\\A")
{'c:\\tmp\\A': StatResult(st_dev=2195292832L, st_ino=6473924464939757L, st_mode=16895, st_mtime=1560413735L),
'c:\\tmp\\A\\B': StatResult(st_dev=2195292832L, st_ino=15199648742969916L, st_mode=16895, st_mtime=1560413742L), 
'c:\\tmp\\A\\B\\C': StatResult(st_dev=2195292832L, st_ino=5348024558097312L, st_mode=16895, st_mtime=1560413739L),}
>>> # Disable access to C:\tmp\A\B
>>> dirsnapshot.DirectorySnapshot("c:\\tmp\\A")
{'c:\\tmp\\A': StatResult(st_dev=2195292832L, st_ino=6473924464939757L, st_mode=16895, st_mtime=1560413735L)}
```

Python 3.6
```python
>>> from watchdog.utils import dirsnapshot
>>> dirsnapshot.DirectorySnapshot("c:\\tmp\\A")
{'c:\\tmp\\A': os.stat_result(st_mode=16895, st_ino=6473924464939757, st_dev=2195292832, st_nlink=1, st_uid=0, st_gid=0, st_size=0, st_atime=1560413735, st_mtime=1560413735, st_ctime=1560411735),
'c:\\tmp\\A\\B': os.stat_result(st_mode=16895, st_ino=15199648742969916, st_dev=2195292832, st_nlink=1, st_uid=0, st_gid=0, st_size=0, st_atime=1560413742, st_mtime=1560413742, st_ctime=1560413306), 
'c:\\tmp\\A\\B\\C': os.stat_result(st_mode=16895, st_ino=5348024558097312, st_dev=2195292832, st_nlink=1, st_uid=0, st_gid=0, st_size=0, st_atime=1560413739, st_mtime=1560413739, st_ctime=1560413739)}
>>> # Disable access to C:\tmp\A\B
>>> dirsnapshot.DirectorySnapshot("c:\\tmp\\A")
{'c:\\tmp\\A': os.stat_result(st_mode=16895, st_ino=6473924464939757, st_dev=2195292832, st_nlink=1, st_uid=0, st_gid=0, st_size=0, st_atime=1560413735, st_mtime=1560413735, st_ctime=1560411735),
'c:\\tmp\\A\\B': os.stat_result(st_mode=16895, st_ino=0, st_dev=0, st_nlink=0, st_uid=0, st_gid=0, st_size=0, st_atime=1560413742, st_mtime=1560413742, st_ctime=1560413306)}
```

In both cases we not get content of B directory, without exceptions, this is great and I think #408 may be closed after merging this patch. But on Python 3.6 snapshot of directory A contains directory B. I think this is just Python 2 specificity, and just need to keep in mind.

File: src/watchdog/utils/dirsnapshot.py (hunk 1 above)
Comment (BoboTiG) [reply src/watchdog/utils/dirsnapshot.py]: Yes, my branch is not up-to-date with master I guess. Will rebase it.

**Comment by BoboTiG** [comment]
This is great, thank you very much for the review and test :muscle:

**Comment by BoboTiG** [comment]
I should add a test, this is somewhat crucial :)

**Comment by BoboTiG** [comment]
Tests added + some refactoring.

**Comment by BoboTiG** [comment]
I will merge if you are still OK @rrzaripov.

**Comment by rrzaripov** [comment]
@BoboTiG I'm checkout latest version of branch, test passed, all looks good for merging!

**Comment by joshuaskelly** [comment]
@BoboTiG Thanks for driving this getting it merged back in! 🎉🏆

**Comment by BoboTiG** [comment]
Thansk to you for the original patch ;)

**Pull request merged**